### ISSUE TITLE
'title' with laststatus=0 causes visual glitches

### ISSUE THREAD

**Issue by jeromecst** [neovim/neovim#18573]
### Neovim version (nvim -v)

v0.6.1

### Vim (not Nvim) behaves the same?

no

### Operating system/version

Gentoo

### Terminal name/version

st 0.8.4

### $TERM environment variable

xterm-256color

### Installation

Portage

### How to reproduce the issue

nvim --clean
:set title
:set laststatus=0

write or edit the file, the title should appear at the wrong place

### Expected behavior

No title displayed in the screen when editing

### Actual behavior

video demo: [![asciicast](https://asciinema.org/a/pfOvNjxfYopuOYRzB1i5HHNvT.svg)](https://asciinema.org/a/pfOvNjxfYopuOYRzB1i5HHNvT)

**Comment by cbirchinger** [comment]
Getting hit since yesterday by glitches caused by "set title" on Gentoo too. Since there was an update of ncurses yesterday from sys-libs/ncurses-6.3_p20211106 to sys-libs/ncurses-6.3_p20220423, i highly suspect some compatibility issues here. The problem shows up with all Terminals (incl. good old XTerm) that I've tested.

Using a TERM from some external sources like Alactritty (TERM=alacritty) will not cause the issue. Only with a terminfo such as "xterm-256color" or "xterm" from ncurses itself.

No other terminal program i use seems to have problems with setting the terminal title though. This includes traditional VIM.

**Comment by justinmk** [comment]
> No other terminal program i use seems to have problems with setting the terminal title though. This includes traditional VIM.

Nvim uses the "title stacking" feature of terminals, Vim doesn't afaik. https://stackoverflow.com/a/20048874/152142

**Comment by cbirchinger** [comment]
Maybe something changed recently in the terminfo for xterm, xterm-256color etc. of ncurses.
There could be new "features" hitting terminfo, that haven't reached other distributions yet.
The ncurses updated in Gentoo was caused by some security bug, so probably with less testing.

On recent Gentoo:
~$ tput -T xterm-256color tsl
%p1%d`%
~$

On Ubuntu 22.04:
~$ tput -T xterm-256color tsl
~$

On Gentoo:
~ % infocmp -L xterm  | grep status
hass_status_line, move_insert_mode, move_standout_mode,
no_pad_char, prtr_silent, status_line_esc_ok,
delete_line=\E[M, dis_status_line=\E[1$~,
from_status_line=\E[0$},
to_status_line=\E[2$~\E[1$}\E[%i%p1%d`,

On Ubuntu 22.04:
~ $ infocmp -L xterm  | grep status
~ $

**Comment by justinmk** [comment]
Maybe https://github.com/neovim/neovim/pull/18570 helps ?

**Comment by cbirchinger** [comment]
I pulled https://github.com/neovim/neovim/commit/baecd359e3c65baa3918966de75f1b91d5953d67.patch and applied it on the neovim 7.0 release but the title still doesn't get set but written somewhere on screen with TERM=xterm-256color or TERM=xterm.

TERM=xterm-256color example with vanilla config:

![nvim2](https://user-images.githubusercontent.com/4973300/168493686-54785c57-f60d-46e8-aeef-1ea6f552019e.gif)

And here is one with TERM=xterm-sl lacking "tput tsl" showing it still working:

![nvim3](https://user-images.githubusercontent.com/4973300/168493823-19ff98b0-5ce4-4962-b873-6aed4af7e89c.gif)

Screenshots are from  VTE but i've verified with plain old Xorg XTerm which is probably the reference implementation for TERM=xterm or TERM=xterm-256color

**Comment by jeromecst** [comment]
I can reproduce that exact same behavior with St.

Setting `TERM=screen-256color` or `TERM=screen` fixes the problem as well.

**Comment by cbirchinger** [comment]
This is the change of the xterm-256color terminfos between the working and the broken version:

old = ncurses-6.3_p20211106
new = ncurses-6.3_p20220423

```
--- a.txt	2022-05-16 00:22:21.904909266 +0200
+++ b.txt	2022-05-16 00:22:30.524002948 +0200
@@ -1,6 +1,6 @@
-#	Reconstructed via infocmp from file: old/etc/terminfo//x/xterm-256color
+#	Reconstructed via infocmp from file: new/etc/terminfo//x/xterm-256color
 xterm-256color|xterm with 256 colors,
-	am, bce, ccc, km, mc5i, mir, msgr, npc, xenl,
+	am, bce, ccc, eslok, hs, km, mc5i, mir, msgr, npc, xenl,
 	colors#0x100, cols#80, it#8, lines#24, pairs#0x10000,
 	acsc=``aaffggiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz{{||}}~~,
 	bel=^G, blink=\E[5m, bold=\E[1m, cbt=\E[Z, civis=\E[?25l,
@@ -9,10 +9,11 @@
 	cud=\E[%p1%dB, cud1=\n, cuf=\E[%p1%dC, cuf1=\E[C,
 	cup=\E[%i%p1%d;%p2%dH, cuu=\E[%p1%dA, cuu1=\E[A,
 	cvvis=\E[?12;25h, dch=\E[%p1%dP, dch1=\E[P, dim=\E[2m,
-	dl=\E[%p1%dM, dl1=\E[M, ech=\E[%p1%dX, ed=\E[J, el=\E[K,
-	el1=\E[1K, flash=\E[?5h$<100/>\E[?5l, home=\E[H,
-	hpa=\E[%i%p1%dG, ht=^I, hts=\EH, ich=\E[%p1%d@,
-	il=\E[%p1%dL, il1=\E[L, ind=\n, indn=\E[%p1%dS,
+	dl=\E[%p1%dM, dl1=\E[M, dsl=\E[1$~, ech=\E[%p1%dX, ed=\E[J,
+	el=\E[K, el1=\E[1K, flash=\E[?5h$<100/>\E[?5l,
+	fsl=\E[0$}, home=\E[H, hpa=\E[%i%p1%dG, ht=^I, hts=\EH,
+	ich=\E[%p1%d@, il=\E[%p1%dL, il1=\E[L, ind=\n,
+	indn=\E[%p1%dS,
 	initc=\E]4;%p1%d;rgb:%p2%{255}%*%{1000}%/%2.2X/%p3%{255}%*%{1000}%/%2.2X/%p4%{255}%*%{1000}%/%2.2X\E\\,
 	invis=\E[8m, is2=\E[!p\E[?3;4l\E[4l\E>, kDC=\E[3;2~,
 	kEND=\E[1;2F, kHOM=\E[1;2H, kIC=\E[2;2~, kLFT=\E[1;2D,
@@ -53,5 +54,5 @@
 	smglr=\E[?69h\E[%i%p1%d;%p2%ds,
 	smgrp=\E[?69h\E[%i;%p1%ds, smir=\E[4h, smkx=\E[?1h\E=,
 	smm=\E[?1034h, smso=\E[7m, smul=\E[4m, tbc=\E[3g,
-	u6=\E[%i%d;%dR, u7=\E[6n, u8=\E[?%[;0123456789]c,
-	u9=\E[c, vpa=\E[%i%p1%dd,
+	tsl=\E[2$~\E[1$}\E[%i%p1%d`, u6=\E[%i%d;%dR, u7=\E[6n,
+	u8=\E[?%[;0123456789]c, u9=\E[c, vpa=\E[%i%p1%dd,

```
No idea how something as basic as xterm-256color still gets such changes in 2022

**Comment by cbirchinger** [comment]
Since the original bug report was from a Gentoo user and my setup is too, I've created a small workaround for Gentoo until a better solution is found:

Put it into /etc/portage/patches/app-editors/neovim-0.7.0/neovim-0.7.0-force-tsl-override.patch and emerge Neovim.
I pinned it's application only to 0.7.0 so future updates don't try and probably fail applying the patch. I haven't tested
it with version 0.6.1 (original report) though. To try replace "neovim-0.7.0" with your version.

```
diff -ruN neovim-0.7.0.orig/src/nvim/tui/tui.c neovim-0.7.0/src/nvim/tui/tui.c
--- neovim-0.7.0.orig/src/nvim/tui/tui.c	2022-04-15 13:38:44.000000000 +0200
+++ neovim-0.7.0/src/nvim/tui/tui.c	2022-05-16 09:27:51.542131058 +0200
@@ -1699,8 +1699,8 @@
     // treatable as xterm.
 
     // 2017-04 terminfo.src lacks these.  Xterm-likes have them.
-    unibi_set_if_empty(ut, unibi_to_status_line, "\x1b]0;");
-    unibi_set_if_empty(ut, unibi_from_status_line, "\x07");
+    unibi_set_str(ut, unibi_to_status_line, "\x1b]0;");
+    unibi_set_str(ut, unibi_from_status_line, "\x07");
     unibi_set_if_empty(ut, unibi_set_tb_margin, "\x1b[%i%p1%d;%p2%dr");
     unibi_set_if_empty(ut, unibi_enter_italics_mode, "\x1b[3m");
     unibi_set_if_empty(ut, unibi_exit_italics_mode, "\x1b[23m");

```
[neovim-0.7.0-force-tsl-override.patch.gz](https://github.com/neovim/neovim/files/8698068/neovim-0.7.0-force-tsl-override.patch.gz)

**Comment by shibotto** [comment]
This is the patch that introduced the issue: https://github.com/mirror/ncurses/commit/8bf8c836fe2c28238adbd62e71f03272efe155ad
It adds [DEC statusline](https://raw.githubusercontent.com/mirror/ncurses/47c323416bb23200896a311ceadbea794b0eb3a6/misc/terminfo.src) (description and definition at L:5600) to xterm-new, which is inherited by xterm and xterm-256color. For now I'm just patching that out, for Gentoo users:

```
diff --git a/misc/terminfo.src b/misc/terminfo.src
index 6a78077..f087b5a 100644
--- a/misc/terminfo.src
+++ b/misc/terminfo.src
@@ -4886,7 +4886,7 @@ xterm-p370|xterm patch #370,
 
 # This version reflects the current xterm features.
 xterm-new|modern xterm terminal emulator,
-	use=dec+sl, use=xterm-p370,
+	use=xterm-p370,
 
 # This fragment is for people who cannot agree on what the backspace key
 # should send.
```

I honestly don't know what the best course of action would be; report it to ncurses since there are a lot of terminal using xterm and xterm-256color that don't support that? Should terminal emulators switch to something else than xterm-256color instead?

**Comment by cbirchinger** [comment]
"xterm-256color" has become the lowest common denominator, because the idea that every new terminal project can expect people to install new terminfo on all the system they maintain is very unrealistic. "xterm-256color" should have been frozen in time years ago. After that maybe focus on something similar to $COLORTERM for more modern features like images (OSC1337) or clipboard (OSC52) support etc. Insisting that xterm-256color is only determined by how the original XTerm behaves, will cause a lot of problems, even though it's technically correct. It will lead to more advanced terminal programs like NVIM simply ignoring terminfo and hard code values or let the user do it.

**Comment by cbirchinger** [comment]
To stay on track, "set title" with this recent ncurses terminfo for xterm-256color is also broken with XTerm 372, which is the reference implementation for that terminfo. So i think there's still an actual bug within Neovim and not just a compatibility quirk it has to deal with.

**Comment by shibotto** [comment]
> To stay on track, "set title" with this recent ncurses terminfo for xterm-256color is also broken with XTerm 372, which is the reference implementation for that terminfo. So i think there's still an actual bug within Neovim and not just a compatibility quirk it has to deal with.

That's because it's opt-in with `--enable-status-line`, which Gentoo (nor any other distro I've checked) sets at build time. That said I don't know how it's supposed to look when enabled, so it might not work for Neovim anyway.

**Comment by pinpox** [comment]
I can reproduce this on other terminal emulators with `xtrem-256color` is set as $TERM.

**Comment by justinmk** [comment]
@pinpox that comment doesn't help without mentioning exact versions / system details.

**Comment by pinpox** [comment]
> @pinpox that comment doesn't help without mentioning exact versions / system details.
```
NVIM v0.7.0
Build type: Release
LuaJIT 2.1.0-beta3
Compiled by nixbld

Features: +acl +iconv +tui
```
Running on NixOS. I tried Wezterm and Alacritty

**Comment by aktau** [comment]
The problems may have been introduced by this ncurses commit, changing the terminfo: https://github.com/mirror/ncurses/commit/8f6d94b8d6211a2323caef53fa4c96c475ec9a6.

**Comment by BoltsJ** [comment]
After upgrading to fedora 37, I'm also having this problem, even with laststatus=2

Fedora 37 (Not working):
```
~:$ toolbox enter fedora-toolbox-37
~:$ nvim --version
NVIM v0.7.2
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
…
~:$ tput -T xterm-256color tsl
%p1%d`~:$  
logout
```
Fedora 36 (Working):
```
~:$ toolbox enter fedora-toolbox-36
~:$ nvim --version
NVIM v0.7.2
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
…
~:$ tput -T xterm-256color tsl
~:$ 
logout
```
On F36 ncurses was `ncurses-base-6.2-9.20210508.fc36.noarch` on 37 it's `ncurses-base-6.3-3.20220501.fc37.noarch`

**Comment by evenreven** [comment]
Just a tiny heads-up for people googling this bug: Fedora 37 apparently ships with a recent version of ncurses, so a standard neovim with `set title` will glitch in the same way for all Fedora beta users (and very soon a lot more users when 37 is released). (Took me a long time and fruitless disabling of plugins to figure this out.)

**Comment by lyokha** [comment]
I have switched from *mate-terminal* to *kitty* because of this issue, and I'm happy.  :smile:  Not only there are no title glitches in the latter, but also it features various cool things not available in other terminal emulators.

**Comment by yut23** [comment]
It looks like the relevant terminfo changes were effectively reverted in https://github.com/mirror/ncurses/commit/7104baa59ff713e8339b3b20edcf4234a5af6200, which is included in ncurses 6.4. See the last bullet point in the [terminal database](https://invisible-island.net/ncurses/announce.html#h3-database) section of the release notes.
